pkg/seedadmission/pods.go

```diff
@@ -0,0 +1,132 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seedadmission
+
+import (
+	"context"
+	"fmt"
+
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardenlogger "github.com/gardener/gardener/pkg/logger"
+
+	"github.com/sirupsen/logrus"
+	"gomodules.xyz/jsonpatch/v2"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// MutatePod mutates pod if mutation conditions are met
+func MutatePod(ctx context.Context, c client.Client, logger *logrus.Logger, request *admissionv1beta1.AdmissionRequest) ([]jsonpatch.JsonPatchOperation, error) {
+	// Ignore all resources other than our expected ones
+	switch request.Resource {
+	case
+		metav1.GroupVersionResource{Group: corev1.SchemeGroupVersion.Group, Version: corev1.SchemeGroupVersion.Version, Resource: "pods"}:
+	default:
+		return nil, nil
+	}
+
+	obj, err := getRequestObject(ctx, c, request)
+	if err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	entryLogger := gardenlogger.
+		NewFieldLogger(logger, "resource", fmt.Sprintf("%s/%s/%s", request.Kind.Group, request.Kind.Version, request.Kind.Kind)).
+		WithField("operation", request.Operation).
+		WithField("namespace", request.Namespace)
+
+	entryLogger.Info("Handling request")
+	patch, err := mutateShootControlPlanePodAnnotations(ctx, c, entryLogger, obj, request.Namespace)
+	if err != nil {
+		logger.Errorf("Error mutating pod %s/%s: %v", request.Name, request.Namespace, err)
+		return nil, err
+	}
+
+	return patch, nil
+}
+
+// mutateShootControlPlanePodAnnotations checks if the object mutation is allowed and return the needed patches .
+func mutateShootControlPlanePodAnnotations(ctx context.Context, c client.Client, logger *logrus.Entry, pod runtime.Object, namespace string) ([]jsonpatch.JsonPatchOperation, error) {
+	var patch []jsonpatch.JsonPatchOperation
+	// determine whether to perform mutation
+	isMutationRequired, err := mutationRequired(ctx, c, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	acc, err := meta.Accessor(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	if !isMutationRequired {
+		logger.Infof("Skipping mutation for %s/%s due to policy check", acc.GetNamespace(), acc.GetName())
```

Comment:
Should we move this to debug log level? I guess a lot of pods are created in the seed for which the mutating is not required. This will flood the logs unnecessarily. WDYT?